tests/distributions/test_transforms.py

```diff
@@ -81,208 +83,97 @@ def _test_shape(self, base_shape, transform):
+        self._test(transform_factory, **kwargs)
+
+    # Affine autoregressive
+    def test_affine_autoregressive(self):
+        for stable in [True, False]:
+            self._test(partial(T.affine_autoregressive, stable=stable))
+
+    # Affine Coupling
+    def test_affine_coupling(self):
+        self._test(T.affine_coupling)
+
+    # Batchnorm
+    def test_batchnorm(self):
+        def transform_factory(input_dim):
             transform = T.batchnorm(input_dim)
             transform._inverse(torch.normal(torch.arange(0., input_dim), torch.arange(1., 1. + input_dim) / input_dim))
             transform.eval()
-            self._test_jacobian(input_dim, transform)
+            return transform
+
+        self._test(transform_factory)
 
+    # Block autoregresive
     def test_block_autoregressive_jacobians(self):
         for activation in ['ELU', 'LeakyReLU', 'sigmoid', 'tanh']:
-            for input_dim in [2, 5, 10]:
-                self._test_jacobian(
-                    input_dim,
-                    T.block_autoregressive(input_dim, activation=activation))
+            self._test(partial(T.block_autoregressive, activation=activation), inverse=False)
 
         for residual in [None, 'normal', 'gated']:
-            for input_dim in [2, 5, 10]:
-                self._test_jacobian(
-                    input_dim,
-                    T.block_autoregressive(input_dim, residual=residual))
-
-    def test_affine_autoregressive_jacobians(self):
-        for stable in [True, False]:
-            for input_dim in [2, 5, 10]:
-                self._test_jacobian(input_dim, T.affine_autoregressive(input_dim, stable=stable))
-
-    def test_tanh_jacobians(self):
-        for input_dim in [2, 5, 10]:
-            self._test_jacobian(input_dim, T.tanh())
-
-    def test_dct_jacobians(self):
-        for input_dim in [2, 5, 10]:
-            self._test_jacobian(input_dim, T.DiscreteCosineTransform())
+            self._test(partial(T.block_autoregressive, residual=residual), inverse=False)
 
-    def test_neural_autoregressive_jacobians(self):
-        for activation in ['ELU', 'LeakyReLU', 'sigmoid', 'tanh']:
-            for input_dim in [2, 5, 10]:
-                self._test_jacobian(input_dim, T.neural_autoregressive(input_dim, activation=activation))
-
-    def test_planar_jacobians(self):
-        for input_dim in [2, 5, 10]:
-            self._test_jacobian(input_dim, T.planar(input_dim))
+    def test_conditional_planar(self):
+        self._test_conditional(T.conditional_planar, inverse=False)
 
-    def test_cond_planar_jacobians(self):
-        observed_dim = 3
-        for input_dim in [2, 5, 10]:
-            z = torch.rand(observed_dim)
-            transform = T.conditional_planar(input_dim, observed_dim)
-            self._test_jacobian(input_dim, transform.condition(z))
+    def test_discrete_cosine(self):
+        # NOTE: Need following hack since helper function unimplemented
```

Comment:
not sure if i would call these "hacks"?

Reply:
Well, it's much cleaner if the help function is in the relevant transform module. I will add this in another (short) PR!